### ISSUE TITLE
ArgumentException: Option 'name' not supported.

### ISSUE THREAD

**Issue by douglasparker** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1901]
### Steps to reproduce
Connect to a MariaDB database using a similar connection string.

Connection string:
```json
{
	"ConnectionStrings": {
		"DefaultConnection": "DataSource=app.db;Cache=Shared",
		"MidgardUniverse": "DataSource=10.0.0.10;Username=midgard;Password=password;Port=3306;Database=midgard;"
	}
}
```

### The issue
My connection string worked fine with `8.0.0-beta.2` but is broken with each stable release after up to `8.0.2`. I'm not sure what the actual issue is because my connection string doesn't contain a `name` option as described in the exception message.

```
Exception message: ArgumentException: Option 'name' not supported.
Stack trace:
MySqlConnector.MySqlConnectionStringOption.GetOptionForKey(string key) in MySqlConnectionStringBuilder.cs
MySqlConnector.MySqlConnectionStringBuilder.set_Item(string key, object value) in MySqlConnectionStringBuilder.cs
System.Data.Common.DbConnectionStringBuilder.set_ConnectionString(string value)
MySqlConnector.MySqlConnectionStringBuilder..ctor(string connectionString) in MySqlConnectionStringBuilder.cs
Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlConnectionSettings..ctor(string connectionString)
Pomelo.EntityFrameworkCore.MySql.Internal.MySqlOptions.GetConnectionSettings(MySqlOptionsExtension relationalOptions)
Pomelo.EntityFrameworkCore.MySql.Internal.MySqlOptions.Initialize(IDbContextOptions options)
Microsoft.EntityFrameworkCore.Internal.SingletonOptionsInitializer.EnsureInitialized(IServiceProvider serviceProvider, IDbContextOptions options)
Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.<GetOrAdd>g__BuildServiceProvider|4_1(IDbContextOptions _, ValueTuple<ConcurrentDictionary<IDbContextOptions, ValueTuple<IServiceProvider, IDictionary<string, string>>>, IDbContextOptions> arguments)
Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache+<>c.<GetOrAdd>b__4_0(IDbContextOptions contextOptions, ValueTuple<ConcurrentDictionary<IDbContextOptions, ValueTuple<IServiceProvider, IDictionary<string, string>>>, IDbContextOptions> tuples)
System.Collections.Concurrent.ConcurrentDictionary<TKey, TValue>.GetOrAdd<TArg>(TKey key, Func<TKey, TArg, TValue> valueFactory, TArg factoryArgument)
Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.GetOrAdd(IDbContextOptions options, bool providerRequired)
Microsoft.EntityFrameworkCore.DbContext.get_ContextServices()
Microsoft.EntityFrameworkCore.DbContext.get_Model()
Microsoft.EntityFrameworkCore.Internal.InternalDbSet<TEntity>.get_EntityType()
Microsoft.EntityFrameworkCore.Internal.InternalDbSet<TEntity>.CheckState()
Microsoft.EntityFrameworkCore.Internal.InternalDbSet<TEntity>.get_EntityQueryable()
Microsoft.EntityFrameworkCore.Internal.InternalDbSet<TEntity>.System.Collections.Generic.IAsyncEnumerable<TEntity>.GetAsyncEnumerator(CancellationToken cancellationToken)
System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable<T>.GetAsyncEnumerator()
Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync<TSource>(IQueryable<TSource> source, CancellationToken cancellationToken)
MidgardUniverse.Website.Pages.Database.ItemDatabaseModel.OnGetAsync() in Items.cshtml.cs
+
        List<ItemDb> itemDb = await _context.ItemDbs.ToListAsync();
Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory+GenericTaskHandlerMethod.Convert<T>(object taskAsObject)
Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory+GenericTaskHandlerMethod.Execute(object receiver, object[] arguments)
Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeHandlerMethodAsync()
Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeNextPageFilterAsync()
Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Rethrow(PageHandlerExecutedContext context)
Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeInnerFilterAsync()
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

### Further technical details

MySQL version: mariadb:11.0
Operating system: ubuntu:22.04
Pomelo.EntityFrameworkCore.MySql version: 8.0.2
Microsoft.AspNetCore.App version: net8.0

Other details about my project setup: n/a

**Comment by lauxjpn** [comment]
@douglasparker The connection string you posted works fine.

It is likely that this is not the connection string that is used by Pomelo/MySqlConnector, since the one you are actually using seems to contain the `name` key.

You can verify this by adding the `CustomMySqlOptions` class and the `.ReplaceService<IMySqlOptions, CustomMySqlOptions>()` call from the following sample console app to your own code base:

<details>
<summary>Program.cs</summary>

```csharp
using System;
using System.Diagnostics;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Infrastructure;
using Microsoft.Extensions.Logging;
using Pomelo.EntityFrameworkCore.MySql.Infrastructure.Internal;
using Pomelo.EntityFrameworkCore.MySql.Internal;

namespace IssueConsoleTemplate;

public class IceCream
{
    public int Id { get; set; }
    public string Name { get; set; }
}

// Define your custom service.
public class CustomMySqlOptions : MySqlOptions
{
    public override void Initialize(IDbContextOptions options)
    {
        var mySqlOptions = options.FindExtension<MySqlOptionsExtension>() ?? new MySqlOptionsExtension();

        if (mySqlOptions.Connection is not null)
        {
            Console.WriteLine($"Initialize -> A connection is used with the following connection string: {mySqlOptions.Connection.ConnectionString}");
        }
        else
        {
            Console.WriteLine($"Initialize -> The following connection string is used: {mySqlOptions.ConnectionString}");
        }

        base.Initialize(options);
    }
}

public class Context : DbContext
{
    public DbSet<IceCream> IceCreams { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        if (!optionsBuilder.IsConfigured)
        {
            var connectionString = "DataSource=127.0.0.1;Username=root;Password=;Port=3306;Database=Issue1901";
            var serverVersion = ServerVersion.AutoDetect(connectionString);

            optionsBuilder
                .UseMySql(connectionString, serverVersion)
                .ReplaceService<IMySqlOptions, CustomMySqlOptions>() // <-- Inject the custom service
                .LogTo(Console.WriteLine, LogLevel.Information)
                .EnableSensitiveDataLogging()
                .EnableDetailedErrors();
        }
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<IceCream>(
            entity =>
            {
                entity.HasData(
                    new IceCream { Id = 1, Name = "Vanilla" },
                    new IceCream { Id = 2, Name = "Chocolate" });
            });
    }
}

internal static class Program
{
    private static void Main()
    {
        using var context = new Context();
        
        context.Database.EnsureDeleted();
        context.Database.EnsureCreated();

        var entities = context.IceCreams
            .OrderBy(i => i.Id)
            .ToList();

        Trace.Assert(entities.Count == 2);
        Trace.Assert(entities[0].Name == "Vanilla");
        Trace.Assert(entities[1].Name == "Chocolate");
    }
}
```

</details>

<details>
<summary>Output</summary>

```
Initialize -> The following connection string is used: DataSource=127.0.0.1;Username=root;Password=;Port=3306;Database=Issue1901
warn: 13.04.2024 12:23:44.329 CoreEventId.SensitiveDataLoggingEnabledWarning[10400] (Microsoft.EntityFrameworkCore.Infrastructure) 
      Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data; this mode should only be enabled during development.
info: 13.04.2024 12:23:44.642 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command) 
      Executed DbCommand (22ms) [Parameters=[], CommandType='Text', CommandTimeout='30']                                
      DROP DATABASE `Issue1901`;                                                                                        
info: 13.04.2024 12:23:44.823 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command) 
      Executed DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='30']                                 
      CREATE DATABASE `Issue1901`;                                                                                      
info: 13.04.2024 12:23:44.994 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command) 
      Executed DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='30']                                 
      ALTER DATABASE CHARACTER SET utf8mb4;                                                                             
info: 13.04.2024 12:23:45.028 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command) 
      Executed DbCommand (33ms) [Parameters=[], CommandType='Text', CommandTimeout='30']                                
      CREATE TABLE `IceCreams` (                                                                                        
          `Id` int NOT NULL AUTO_INCREMENT,                                                                             
          `Name` longtext CHARACTER SET utf8mb4 NULL,
          CONSTRAINT `PK_IceCreams` PRIMARY KEY (`Id`)
      ) CHARACTER SET=utf8mb4;
info: 13.04.2024 12:23:45.037 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed DbCommand (8ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      INSERT INTO `IceCreams` (`Id`, `Name`)
      VALUES (1, 'Vanilla'),
      (2, 'Chocolate');
info: 13.04.2024 12:23:45.314 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command) 
      Executed DbCommand (5ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `i`.`Id`, `i`.`Name`
      FROM `IceCreams` AS `i`
      ORDER BY `i`.`Id`
```

</details>

It will output the connection string that you pass to Pomelo.

**Comment by douglasparker** [comment]
> @douglasparker The connection string you posted works fine.
> 
> It is likely that this is not the connection string that is used by Pomelo/MySqlConnector, since the one you are actually using seems to contain the `name` key.

The name key isn’t used in any connection string in the codebase, which is why I am confused.

Reverting to `8.0.0-beta.2 ` allows the specified connection string to work, without any errors. Connection to the database is successful.

Going back to stable results in what I posted above. That tells me that something with this connector has changed and is resulting in the working behavior and broken behavior depending on version.

**Comment by lauxjpn** [comment]
@douglasparker Did you inject the service I posted in my previous comment to output the connection string that is being used by Pomelo? Please post the result of the output, so we can examine it (replace sensitive information). Thanks!

**Comment by douglasparker** [comment]
@lauxjpn sorry for the delayed response!

It points to the same valid connection string! I'm not sure why it behaves properly on `8.0.0-beta.2`, but clearly something changed since then that is causing this issue.

```
Initialize -> The following connection string is used: DataSource=10.0.0.10;Username=midgard;Password=;Port=3306;Database=midgard;
Initialize -> The following connection string is used: name=ConnectionStrings:MidgardUniverse
```

**Comment by lauxjpn** [comment]
> It points to the same valid connection string!

That doesn't seem to be the case. You posted that _two_ outputs where generated. One with the connection string that you expect  and then a _second_ one, that contains `name=ConnectionStrings:MidgardUniverse`.

The second one likely overwrites the first one (or it is using a second `DbContext` instance; it is impossible to tell just from the output).

Since we now found the reason _why_ the exception is being thrown, the next step is to find out _where_ in your codebase the `name=ConnectionStrings:MidgardUniverse` connection string is set.

@douglasparker Do you want to just make a global search for it (or `MidgardUniverse`) over your codebase, or do you want me to post some code that returns a stack trace every time a connection string gets set?

**Comment by douglasparker** [comment]
> > It points to the same valid connection string!
> 
> That doesn't seem to be the case. You posted that _two_ outputs where generated. One with the connection string that you expect and then a _second_ one, that contains `name=ConnectionStrings:MidgardUniverse`.

Ah, it looks like there is an override for some reason in some scaffolded code for a remote database! I will investigate this further in the morning.

I really appreciate your help with this! Very strange that this wasn’t an issue in the beta version- but I am very happy to see that this seems to be user error. 😅

**Comment by lauxjpn** [comment]
> [...] but I am very happy to see that this seems to be user error.

Actually, we don't know yet whether this is just a user error or whether this is a regression bug. Since your code seems to have worked in `8.0.0-beta.2`, we would still want to investigate this a bit further.

Once you found the overriding code, please post the code that sets the original (correct) connection string and the one that then sets the `name=` one (remove/replace sensitive data and post a couple of lines before and after the code in question for context, respectively), so we can reproduce the scenario on our end.

If it turns out to be a regression bug, we should fix it. Thanks!

**Comment by douglasparker** [comment]
**Program.cs**

```
var gameConnectionString = builder.Configuration.GetConnectionString("MidgardUniverse") ?? throw new InvalidOperationException("Connection string 'MidgardUniverse' not found.");
builder.Services.AddDbContext<MidgardContext>(options =>
    options.UseMySql(gameConnectionString, ServerVersion.AutoDetect(gameConnectionString)));
```

**MidgardDbContext.cs** (Scaffolded)

```
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder.UseMySql("name=ConnectionStrings:MidgardUniverse", Microsoft.EntityFrameworkCore.ServerVersion.Parse("11.0.4-mariadb"));

```

**Comment by lauxjpn** [comment]
This is indeed a regression bug, since named connection strings worked in Pomelo `7.0` but don't anymore in `8.0+`.
We will fix and backport it.

**Comment by douglasparker** [comment]
> This is indeed a regression bug, since named connection strings worked in Pomelo `7.0` but don't anymore in `8.0+`. We will publish a fix and backport it to `8.0.x`.

Ah, that’s great news! Thanks for helping me look into this!

**Comment by RoderickEllis** [comment]
I would like to mention that this issue has not been backported to 8.0.x, and I was wondering if that was still being planned. Our projects all target .Net 8 for long-term support, and I wanted to confirm that this bug was going to be fixed before I attempted to move our projects to a potentially new Entity Framework library.
